File: coding-interviewes/coding_interview_150/java/InsertDeleteGetRandomO1.java

/* Insert Delete GetRandom O(1) — ArrayList+HashMap swap-remove. */
import java.util.*;
public class InsertDeleteGetRandomO1 {
    static class RandomizedSet {
        List<Integer> arr=new ArrayList<>(); Map<Integer,Integer> pos=new HashMap<>(); Random rnd=new Random();
        public boolean insert(int v){ if (pos.containsKey(v)) return false; pos.put(v,arr.size()); arr.add(v); return true; }
        public boolean remove(int v){ Integer i=pos.get(v); if (i==null) return false; int last=arr.get(arr.size()-1);
            arr.set(i,last); pos.put(last,i); arr.remove(arr.size()-1); pos.remove(v); return true; }
        public int getRandom(){ return arr.get(rnd.nextInt(arr.size())); }
    }
    public static void main(String[] args){
        RandomizedSet rs=new RandomizedSet(); System.out.println(rs.insert(1)); System.out.println(rs.remove(2)); System.out.println(rs.insert(2)); System.out.println(rs.getRandom());
    }
}

File: coding-interviewes/coding_interview_150/scala/InsertDeleteGetRandomO1.scala

/* Insert Delete GetRandom O(1) — ArrayBuffer+HashMap */
import scala.util.Random
import scala.collection.mutable
object InsertDeleteGetRandomO1 {
  class RandomizedSet {
    val arr = new scala.collection.mutable.ArrayBuffer[Int]()
    val pos = mutable.HashMap[Int,Int]()
    val rnd = new Random()
    def insert(v:Int):Boolean = {
      if (pos.contains(v)) return false
      pos(v) = arr.length; arr += v; true
    }
    def remove(v:Int):Boolean = pos.get(v) match {
      case None => false
      case Some(i) =>
        val last = arr(arr.length-1)
        arr(i) = last; pos(last) = i
        arr.remove(arr.length-1); pos.remove(v); true
    }
    def getRandom():Int = arr(rnd.nextInt(arr.length))
  }
  def main(args:Array[String]):Unit = {
    val rs = new RandomizedSet
    println(rs.insert(1)); println(rs.remove(2)); println(rs.insert(2)); println(Set(1,2).contains(rs.getRandom()))
  }
}
